### PR TITLE
Ensure only measured subcircuits cause full simulator sweeps

### PR DESCRIPTION
**Pull request by daxfohl**
Fixes a perf issue that *any* subcircuit would cause full sweep/repeat, even if it was unmeasured. With this change, at least only measured subcircuits will cause full sweep. (It does not however optimize for the case where such a subcircuit itself only contains terminal measurements, which is a deeper problem).

This PR also generalizes `split_into_unitary_then_general` to allow for splitting by any protocol.

@95-martin-orion

### CODE REVIEW

**Review by 95-martin-orion — APPROVED** [review]
Thanks for following up on this! In a future PR, it may be possible to combine the circuit-splitting method here with `mapped_circuit` from #3785 to recursively split `CircuitOperation`s as well.

File: cirq/sim/density_matrix_simulator.py [hunk 1]
```diff
@@ -168,8 +168,11 @@ def _run(
         resolved_circuit = protocols.resolve_parameters(circuit, param_resolver)
         check_all_resolved(resolved_circuit)
 
-        if circuit.are_all_measurements_terminal() and not any(
-            circuit.findall_operations(lambda op: isinstance(op, circuits.CircuitOperation))
+        unmeasured_prefix, general_suffix = split_into_matching_protocol_then_general(
+            resolved_circuit, lambda op: not protocols.is_measurement(op)
+        )
+        if len(general_suffix) <= 1 and not any(
+            general_suffix.findall_operations(lambda op: isinstance(op, circuits.CircuitOperation))
```
Comment (95-martin-orion) [inline cirq/sim/density_matrix_simulator.py:175]: Almost missed this: applying the previous `circuit`-based check to `general_suffix` should work here. `general_circuit` can be longer than one moment as long as all measurements are terminal and there are no `CircuitOperation`s.

**Review by 95-martin-orion — CHANGES_REQUESTED** [review]
Sorry, caught something at the last second.

**Comment by daxfohl** [comment]
Will change it when I get back, but actually don't see how it would happen,
given the "split" implementation.

File: cirq/sim/density_matrix_simulator.py (hunk 1 above)
Comment (daxfohl) [reply cirq/sim/density_matrix_simulator.py]: fixed

**Review by 95-martin-orion — APPROVED** [review]

**Pull request merged**